packages/mobile/src/merchantPayment/MerchantPaymentScreen.tsx

```diff
@@ -0,0 +1,286 @@
+import { ContractKit } from '@celo/contractkit'
+import { Charge, ContractKitTransactionHandler } from '@celo/payments-sdk'
+import { AbortCodes, BusinessData } from '@celo/payments-types'
+import ReviewFrame from '@celo/react-components/components/ReviewFrame'
+import colors from '@celo/react-components/styles/colors'
+import fontStyles from '@celo/react-components/styles/fonts'
+import { UnlockableWallet } from '@celo/wallet-base'
+import { useNavigation } from '@react-navigation/native'
+import { StackScreenProps } from '@react-navigation/stack'
+import BigNumber from 'bignumber.js'
+import React, { useCallback, useEffect, useState } from 'react'
+import { useTranslation } from 'react-i18next'
+import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
+import { SafeAreaView } from 'react-native-safe-area-context'
+import { useDispatch, useSelector } from 'react-redux'
+import { e164NumberSelector } from 'src/account/selectors'
+import { showError, showMessage } from 'src/alert/actions'
+import ContactCircle from 'src/components/ContactCircle'
+import FeeDrawer from 'src/components/FeeDrawer'
+import TokenDisplay from 'src/components/TokenDisplay'
+import TokenTotalLineItem from 'src/components/TokenTotalLineItem'
+import { estimateFee, FeeType } from 'src/fees/reducer'
+import { feeEstimatesSelector } from 'src/fees/selectors'
+import { UNLOCK_DURATION } from 'src/geth/consts'
+import { useCurrencyToLocalAmount } from 'src/localCurrency/hooks'
+import { getLocalCurrencyCode } from 'src/localCurrency/selectors'
+import { navigateBack, navigateHome } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { StackParamList } from 'src/navigator/types'
+import { getPassword, retrieveOrGeneratePepper } from 'src/pincode/authentication'
+import { useTokenInfoBySymbol } from 'src/tokens/hooks'
+import { fetchTokenBalances } from 'src/tokens/reducer'
+import { Currency } from 'src/utils/currencies'
+import { divideByWei } from 'src/utils/formatting'
+import { getContractKitAsync } from 'src/web3/contracts'
+import { isDekRegisteredSelector } from 'src/web3/selectors'
+
+type Props = StackScreenProps<StackParamList, Screens.MerchantPayment>
+
+function MerchantPaymentScreen({ route }: Props) {
```

Comment:
I don't want to be annoying because I know there's a time crunch here, but it feels like this screen is doing too much and some things can be sent to other helper components or custom hooks.

For example:
- All the fee handling should probably be in it's own component.
- All the payments stuff and interaction with the payments SDK could be a hook that holds all the logic related to it and just return whatever is needed to draw the UI.

So basically anything that is encapsulated or pure logic should be outside of the screen and just UI-related things should be in this file.

I know not all files in this repo follow the above though, there are some screens that are not properly designed but we do like to avoid it in new components.

What do you think @jeanregisser ?

Reply:
Yes exactly @gnardini.

Adding to this, this merchant payment flow should be probably be handled by a saga, taking into account errors and cancellation at any point.
There's an example of this with Bidali in https://github.com/valora-inc/wallet/blob/6d8719ed101ee48a9bd0548b271803c5484e1251/packages/mobile/src/fiatExchanges/saga.ts#L69-L103

Of course, I realize it's gonna be difficult to make these changes now and meet the deadline.

Looks like we need to chat directly about this to be aligned.